### PR TITLE
Add duplicates and tube_ids feature

### PR DESCRIPTION
**Pull request by CatFish47**
Addresses issue #3275 

- Adds a duplicate column formatted like "SAMPLE_NAME × 2" if there are 2 copies of "SAMPLE_NAME"
- Removed all duplicates from all other columns (blanks, matching, extra)
- Added checking for tube_ids. If tube_ids are detected, will validate on tube_ids and will add the tube id after the sample name in listen (e.g. "SAMPLE_NAME (tube_id)")

### CODE REVIEW

File: qiita_pet/handlers/admin_processing_job.py [hunk 1]
```diff
@@ -105,28 +105,70 @@ def get(self):
...
     @execute_as_transaction
     def post(self):
-
-        # Get user-inputted qiita id and sample names
+        # Get user-inputted qiita id and sample names / tube_ids
         qid = self.get_argument("qid")
         snames = self.get_argument("snames").split()
 
+        # Get study give qiita id
+        st = Study(qid).sample_template
```
Comment (charles-cowart) [inline qiita_pet/handlers/admin_processing_job.py:113]: st is a not a properly descriptive name for a variable, especially one that plays an important role some ten lines later in the code.

**Review by charles-cowart — CHANGES_REQUESTED** [review]
Please consider my comments and add a unit-test for the functionality that you wrote. The test should include BLANKS in upper and lower-case as well as normal sample-names. It should also include sample-names with the qiita-id pre-pended and not.

File: qiita_pet/handlers/admin_processing_job.py [hunk 2]
```diff
@@ -105,28 +105,70 @@ def get(self):
...
         # Stripping leading qiita id from sample names
         # Example: 1.SKB1.640202 -> SKB1.640202
-        qsnames = list(Study(qid).sample_template)
+        qsnames = list(st)
         for i, qsname in enumerate(qsnames):
             if qsname.startswith(qid):
                 qsnames[i] = qsname.replace(f'{qid}.', "", 1)
 
+        # Adds tube ids to a dict with key as tube id and value as qsname
+        tube_ids_dict = dict()
```
[the remark this answers is not in the document]
Comment (CatFish47) [reply qiita_pet/handlers/admin_processing_job.py:123]: In this case, would `tube_id_lookup` be an appropriate name for the variable? Would it be better if I also change the name of the `tube_id_rev` variable too as well then?

File: qiita_pet/handlers/admin_processing_job.py [hunk 3]
```diff
@@ -105,28 +105,70 @@ def get(self):
...
+
         # Stripping leading qiita id from sample names
         # Example: 1.SKB1.640202 -> SKB1.640202
-        qsnames = list(Study(qid).sample_template)
+        qsnames = list(st)
         for i, qsname in enumerate(qsnames):
             if qsname.startswith(qid):
                 qsnames[i] = qsname.replace(f'{qid}.', "", 1)
 
+        # Adds tube ids to a dict with key as tube id and value as qsname
```
[the remark this answers is not in the document]
Comment (CatFish47) [reply qiita_pet/handlers/admin_processing_job.py:122]: Would this be appropriate:
```
# Creates a way to access a tube_id by its corresponding sample name
# and vice versa, which is important to adding tube_id in parentheses
# after a sample name a few lines later
```

File: qiita_pet/handlers/admin_processing_job.py (hunk 1 above)
Comment (CatFish47) [reply qiita_pet/handlers/admin_processing_job.py:113]: Would changing the variable name to `study` work better, or would a variable name like `qt_study` be more appropriate?

**Comment by CatFish47** [comment]
Thanks for the comments! I've pushed by current changes based on the comments that are there if you would like to take another look. Quick question -- how and where do I implement unit tests?

File: qiita_pet/handlers/admin_processing_job.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (charles-cowart) [reply qiita_pet/handlers/admin_processing_job.py]: Sure, those sound good.

File: qiita_pet/handlers/admin_processing_job.py (hunk 1 above)
Comment (charles-cowart) [reply qiita_pet/handlers/admin_processing_job.py]: study is fine.

File: qiita_pet/handlers/admin_processing_job.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (charles-cowart) [reply qiita_pet/handlers/admin_processing_job.py]: much improved, thanks!

**Comment by charles-cowart** [comment]
@CatFish47 Try making a tests directory under qiita_pet/handlers and use the test framework in analysis_handlers/tests and api_proxy/tests as a template. Try to exercise this code.

**Comment by antgonza** [comment]
Closing as this has been superseded by https://github.com/qiita-spots/qiita/pull/3295

**Pull request closed without merging**